Optimize TracingInvocationEventHandler for nested use

carterkozak opened the pull request:
Prevent trace name calculation for unsampled spans.
Longer term we should proivide a better startSpan API
which either takes a string supplier, or uses slf4j style
interpolation.

Before:
```
Benchmark                                     Mode  Cnt    Score    Error  Units
ProxyBenchmark.instrumentedWithTracingNested  avgt    5  900.542 ± 29.627  ns/op
```

After:
```
Benchmark                                     Mode  Cnt    Score   Error  Units
ProxyBenchmark.instrumentedWithTracingNested  avgt    5  214.656 ± 7.242  ns/op
```

## After this PR
==COMMIT_MSG==
Optimize TracingInvocationEventHandler for nested use
==COMMIT_MSG==

carterkozak committed:
Use Blackhole instead of AsyncSlf4jSpanObserver in benchmarks

carterkozak committed:
Optimize TracingInvocationEventHandler for nested usage

Prevent trace name calculation for unsampled spans.
Longer term we should proivide a better startSpan API
which either takes a string supplier, or uses slf4j style
interpolation.

Before:
```
Benchmark                                     Mode  Cnt    Score    Error  Units
ProxyBenchmark.instrumentedWithTracingNested  avgt    5  900.542 ± 29.627  ns/op
```

After:
```
Benchmark                                     Mode  Cnt    Score   Error  Units
ProxyBenchmark.instrumentedWithTracingNested  avgt    5  214.656 ± 7.242  ns/op
```

carterkozak committed:
Add generated changelog entries

schlosna commented on tritium-tracing/src/main/java/com/palantir/tritium/tracing/TracingInvocationEventHandler.java:
```diff
@@ -90,6 +90,15 @@ private void complete(@Nullable InvocationContext context) {
+     * if not present in order to create a new traceId regardless of sampling state to preserve behavior.
+     */
+    static boolean inUnsampledTrace() {
+        return Tracer.hasTraceId() && !Tracer.isTraceObservable();
```
probably overkill optimization, but could push this upstream to `Tracer` to avoid the double thread local lookup:

```
    public static boolean isUnsampledTrace() {
        Trace trace = currentTrace.get();
        return trace != null && !trace.isObservable();
    }
```

carterkozak replied:
Longer term I'd prefer to add an API to take care of the expensive piece, so we don't need to think about this, e.g.

```java
// lambda allocation, but the API is easy to use
Tracer.fastStartSpan(() -> "foo" + bar);
```
or
```java
// garbage free, but requires many overloads.
Tracer.fastStartSpan("foo {}", bar);
```

carterkozak committed:
Optimize TaggedMetricsServiceInvocationEventHandler (#358)

Create the global failure meter eagerly, cache timers by method
to avoid recreating metric name objects.

Before:
```
Benchmark                                     Mode  Cnt    Score     Error  Units
ProxyBenchmark.instrumentedWithTaggedMetrics  avgt    5  669.303 ± 108.389  ns/op
```

After:
```
Benchmark                                     Mode  Cnt    Score    Error  Units
ProxyBenchmark.instrumentedWithTaggedMetrics  avgt    5  233.019 ± 49.507  ns/op
```
